### PR TITLE
ComposeActivity refactor

### PR DESCRIPTION
**Pull request by charlag**
Compose Refactoring - TODOs
   - [X] Save statuses
  CLOSED: [2019-10-20 Sun 21:55]
   - [X] Poll view
  CLOSED: [2019-10-23 Wed 21:21]
   - [X] Poll data
  CLOSED: [2019-10-23 Wed 21:21]
   - [ ] Check initialization from redraft 
   - [ ] Scheduled statuses
   - [ ] Tests

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeActivity.kt [hunk 1]
```diff
@@ -0,0 +1,1171 @@
...
+                instanceParams.pollMaxLength)
+    }
+
+    fun updatePoll(poll: NewPoll) {
+        enableButton(composeAddMediaButton, false, false)
+
+        if (pollPreview == null) {
+            pollPreview = PollPreviewView(this)
+
+            val resources = resources
```
[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeActivity.kt]: Yup, needs cleanup

File: app/src/main/java/com/keylesspalace/tusky/AccountActivity.kt [hunk 2]
```diff
@@ -674,9 +675,8 @@ class AccountActivity : BottomSheetActivity(), ActionButtonActivity, HasAndroidI
 
     private fun mention() {
         loadedAccount?.let {
-            val intent = ComposeActivity.IntentBuilder()
-                    .mentionedUsernames(setOf(it.username))
-                    .build(this)
+            val intent = ComposeActivity.startIntent(this,
+                    ComposeActivity.ComposeOptions(mentionedUsernames = listOf(it.username)))
```
[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/AccountActivity.kt]: Is there a set which is good for parcelizing though?

[the remark this answers is not in the document]
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/AccountActivity.kt]: I think the Kotlin Parcelize just handles it transparently? It probably makes a list out of it or something

[the remark this answers is not in the document]
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/AccountActivity.kt]: Indeed it does, I decompiled it and it just writes the size and some strings
```java
         Set var3 = this.mentionedUsernames;
         Iterator var10000;
         if (var3 != null) {
            parcel.writeInt(1);
            parcel.writeInt(var3.size());
            var10000 = var3.iterator();

            while(var10000.hasNext()) {
               parcel.writeString((String)var10000.next());
            }
         } else {
            parcel.writeInt(0);
         }
```
Thanks for the good tip!
What about the order, is it important to us?

**Comment by charlag** [comment]
Bitrise fails to clone it for some reason :thinking:

**Comment by connyduck** [comment]
I think it is merging this branch into develop before it builds it and there are merge conflicts.

**Comment by connyduck** [comment]
please review @charlag

File: app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeViewModel.kt [hunk 3]
```diff
@@ -93,6 +88,35 @@ class ComposeViewModel
 
     private val mediaToDisposable = mutableMapOf<Long, Disposable>()
 
+
+    init {
+
+        Single.zip(api.getCustomEmojis(), api.getInstance(), BiFunction<List<Emoji>, Instance, InstanceEntity> { emojis, instance ->
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeViewModel.kt]: There should be `Singles` or something like that from Kotlin package to not write all these `BiFunction<...>`

File: app/src/test/java/com/keylesspalace/tusky/ComposeActivityTest.kt [hunk 4]
```diff
@@ -119,15 +98,27 @@ class ComposeActivityTest {
...
+                Single.just(InstanceEntity(instanceDomain, emptyList(),null, null, null, null))
+        )
+
         val dbMock = mock(AppDatabase::class.java)
         `when`(dbMock.instanceDao()).thenReturn(instanceDaoMock)
 
-        activity.mastodonApi = apiMock
-        activity.accountManager = accountManagerMock
-        activity.database = dbMock
+        val viewModel = ComposeViewModel(
```
Comment (charlag) [inline app/src/test/java/com/keylesspalace/tusky/ComposeActivityTest.kt]: Idk if it''s a good idea to have a real ViewModel but activity test now is more "integration" one in any case. We probably need some `ComposeViewModel` tests as well but there's barely any logic there so it's not critical.

File: app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeViewModel.kt (hunk 3 above)
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeViewModel.kt]: did not know this, nice

**Pull request merged**